### PR TITLE
Restore key counters state on replay rewinding

### LINKED ISSUE #3080: Rewinding a replay should decrement the key counters
Originally reported in #1825.

image

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/61115875-rewinding-a-replay-should-decrement-the-key-counters?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UselessToucan**
resolves #3080 

I'm trying to resolve #3080 using Memento design pattern.
Participants:
`KeyCounterMemento` - Memento
`KeyCounter` - Originator
`KeyCounterCollection` - Caretaker

The problem is that seeking backwards causes the game to perform key presses which ruins the restored state. ([Video](https://youtu.be/H4S4GAfpfbA))

### CODE REVIEW

**Comment by peppy** [comment]
`Memento` -> `State`, as we call such states everywhere else in the game.

File: osu.Game/Screens/Play/KeyCounterCollection.cs [hunk 1]
```diff
@@ -21,6 +23,8 @@ public class KeyCounterCollection : FillFlowContainer<KeyCounter>
         public readonly Bindable<bool> Visible = new Bindable<bool>(true);
         private readonly Bindable<bool> configVisibility = new Bindable<bool>();
 
+        private readonly Dictionary<string, List<KeyCounterMemento>> keyCountersState = new Dictionary<string, List<KeyCounterMemento>>();
```
Comment (peppy) [inline osu.Game/Screens/Play/KeyCounterCollection.cs]: can't exist (ugly and should not be `string` based). why not store states inside the `KeyCounter` removing the need for this?

File: osu.Game/Screens/Play/KeyCounterMemento.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+namespace osu.Game.Screens.Play
+{
+    public class KeyCounterMemento
```
Comment (peppy) [inline osu.Game/Screens/Play/KeyCounterMemento.cs]: KeyCounterState

File: osu.Game/Screens/Play/KeyCounterMemento.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        public KeyCounterMemento(double currentTime, int countPresses)
+        {
+            CurrentTime = currentTime;
+            CountPresses = countPresses;
+        }
+
+        public double CurrentTime { get; }
```
Comment (peppy) [inline osu.Game/Screens/Play/KeyCounterMemento.cs]: Time

File: osu.Game/Screens/Play/KeyCounterMemento.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+        public int CountPresses { get; }
```
Comment (peppy) [inline osu.Game/Screens/Play/KeyCounterMemento.cs]: Count

File: osu.Game/Screens/Play/KeyCounterMemento.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Screens/Play/KeyCounterMemento.cs]: also these can probably be `readonly`

File: osu.Game/Screens/Play/KeyCounterCollection.cs [hunk 5]
```diff
@@ -36,6 +40,19 @@ public override void Add(KeyCounter key)
...
+
+            keyCountersState.Add(key.Name, new List<KeyCounterMemento>());
+            key.KeyPressed += () => keyCountersState[key.Name].Add(key.SaveState());
+        }
+
+        public void RestoreKeyCounterState(double time)
+        {
+            foreach (var counter in Children)
+            {
+                var targetState = keyCountersState[counter.Name].LastOrDefault(state => state.CurrentTime <= time);
```
Comment (peppy) [inline osu.Game/Screens/Play/KeyCounterCollection.cs]: what happens on default (null)? looks to be nullref exception?

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
also no tests.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/KeyCounter.cs [hunk 6]
```diff
@@ -55,6 +63,8 @@ protected set
         public Color4 KeyUpTextColor { get; set; } = Color4.White;
         public int FadeTime { get; set; }
 
+        public IClock AudioClock { get; set; }
```
[the remark this answers is not in the document]
Comment (UselessToucan) [reply osu.Game/Screens/Play/KeyCounter.cs]: Can't find any `public` clock objects upward of the scene graph.
https://github.com/ppy/osu/blob/master/osu.Game/Screens/Play/Player.cs#L65
https://github.com/ppy/osu/blob/master/osu.Game/Screens/Play/PauseContainer.cs#L43
https://github.com/ppy/osu/blob/master/osu.Game/Screens/Play/PauseContainer.cs#L44

File: osu.Game/Screens/Play/HUDOverlay.cs [hunk 7]
```diff
@@ -82,7 +82,11 @@ public HUDOverlay(ScoreProcessor scoreProcessor, RulesetContainer rulesetContain
             Progress.Objects = rulesetContainer.Objects;
             Progress.AudioClock = offsetClock;
             Progress.AllowSeeking = rulesetContainer.HasReplayLoaded;
-            Progress.OnSeek = pos => adjustableClock.Seek(pos);
+            Progress.OnSeek = pos =>
+            {
+                adjustableClock.Seek(pos);
+                KeyCounter.RestoreKeyCounterState(pos);
```
[the remark this answers is not in the document]
Comment (UselessToucan) [reply osu.Game/Screens/Play/HUDOverlay.cs]: Don't understand where to find `ElapsedTime`
image

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/HUDOverlay.cs]: `Drawable.Clock.ElapsedFrameTime`

File: osu.Game/Screens/Play/KeyCounter.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/KeyCounter.cs]: `this.Clock`

**Review by peppy — CHANGES_REQUESTED** [review]
as mentioned

File: osu.Game/Screens/Play/KeyCounter.cs [hunk 8]
```diff
@@ -152,6 +152,15 @@ protected override void Update()
                 RestoreState(Clock.CurrentTime);
         }
 
-        public void RestoreState(double time) => CountPresses = states.LastOrDefault(state => state.Time <= time)?.Count ?? 0;
+        public void RestoreState(double time)
+        {
+            var targetState = states.LastOrDefault(state => state.Time <= time);
+            var targetIndex = states.IndexOf(targetState);
+
+            states.RemoveRange(targetIndex + 1, states.Count - (targetIndex + 1));
```
Comment (peppy) [inline osu.Game/Screens/Play/KeyCounter.cs]: in the case targetIndex is null you probably don't want to run this.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/KeyCounterCollection.cs [hunk 9]
```diff
@@ -37,6 +38,8 @@ public override void Add(KeyCounter key)
             key.FadeTime = FadeTime;
             key.KeyDownTextColor = KeyDownTextColor;
             key.KeyUpTextColor = KeyUpTextColor;
+            if (AudioClock != null && AudioClock is IFrameBasedClock basedClock)
+                key.Clock = basedClock;
```
Comment (peppy) [inline osu.Game/Screens/Play/KeyCounterCollection.cs]: this needs a comment as to why it is done.

File: osu.Game/Screens/Play/KeyCounterCollection.cs [hunk 10]
```diff
@@ -37,6 +38,8 @@ public override void Add(KeyCounter key)
             key.FadeTime = FadeTime;
             key.KeyDownTextColor = KeyDownTextColor;
             key.KeyUpTextColor = KeyUpTextColor;
+            if (AudioClock != null && AudioClock is IFrameBasedClock basedClock)
```
Comment (peppy) [inline osu.Game/Screens/Play/KeyCounterCollection.cs]: Is there any reason it's not just an `IFrameBasedClock` in the first place?

**Review by peppy — APPROVED** [review]

**Pull request merged**